### PR TITLE
Fully stop RetryableAction when cancelled

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently cancelling the RetryableAction does not stop one last run from
being executed. This commit makes a best effort attempt to cancel a
scheduled retry and guards future executions from the action already
being completed.

### CODE REVIEW

**Comment by tbrooks8** [comment]
Maybe more of this work should be implemented in the `ThreadPool`? Cancelling the task only cancels the execution on the `ScheduledThreadPoolExecutor`. If that task has successfully executed and scheduled the thing on the final executor (say `GENERIC`), the task will still run on `GENERIC` even if it is still queued at cancellation time.

File: server/src/main/java/org/elasticsearch/action/support/RetryableAction.java [hunk 1]
```diff
@@ -114,6 +123,7 @@ private RetryingListener(long delayMillisBound, ArrayDeque<Exception> caughtExce
 
         @Override
         public void onResponse(Response response) {
+            retryTask = null;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/RetryableAction.java:126]: should we also null it in the else branch of RetryingListener.onFailure?

File: server/src/main/java/org/elasticsearch/action/support/RetryableAction.java [hunk 2]
```diff
@@ -75,6 +78,9 @@ public void run() {
 
     public void cancel(Exception e) {
         if (isDone.compareAndSet(false, true)) {
+            if (retryTask != null) {
+                retryTask.cancel();
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/support/RetryableAction.java:82]: Can't this stilll throw an NPE as we have two consecutive volatile reads?

**Review by dnhatn — APPROVED** [review]
LGTM

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**